SumUp - Setup, Scrub and Purchase build

sinourain opened the pull request:
Description
-------------------------
Setup, scrub and purchase method to SumUp Gateway adapter with the basic information needed.

This are the relevant links to review the initial implementation:

- [SumUp REST API](https://developer.sumup.com/docs/api/sum-up-rest-api/) 
- [Make a payment with a card entered by a customer](https://developer.sumup.com/docs/online-payments/guides/single-payment/) 
- [Checkouts](https://developer.sumup.com/docs/api/checkouts/)

Tickets for Spreedly reference
SER-764
SER-712
SER-711

Unit test
-------------------------
Finished in 37.342409 seconds.
5606 tests, 78027 assertions, 0 failures, 0 errors, 0 pendings, 0 omissions, 0 notifications

Remote test
-------------------------
100% passed
150.12 tests/s, 2089.50 assertions/s

Rubocop
-------------------------
Inspecting 769 files
769 files inspected, no offenses detected

Heavyblade commented on lib/active_merchant/billing/gateways/sum_up.rb:
```diff
@@ -0,0 +1,192 @@
+
+      def upsert_checkout(money, payment, options = {})
+        if options[:checkout_id].present?
+          find_checkout(options[:checkout_id])
```
❓ Question:

What is the point of finding a checkout to extract it's ID if you already have the ID on the `options[:checkout_id]` ?

sinourain replied:
Good catch, you're right

Heavyblade commented on lib/active_merchant/billing/gateways/sum_up.rb:
```diff
@@ -0,0 +1,192 @@
+      end
+
+      def add_payment(post, payment, options)
+        post[:payment_type] = options[:payment_type] if PAYMENT_TYPES.include?(options[:payment_type])
```
❓ Question:

Wouldn't be possible to identify the 'payment_type' from the card brand in conjunction with the payment method class ?

sinourain replied:
I looked in the code and I don't see a way to do it, maybe in another instance it could work, not in this gem

Heavyblade replied:
One idea => 

Well we could confirm with Charlotte/Nick but instead of thinking that the merchant is going to send us a GSF with the proper value for `options[:payment_type]` we could just start with the type 'card' of have it as a fallback if no value is available on `options[:payment_type]`

sinourain replied:
Perfect, I'm going to ask and for now I'll leave `card` as the default value

Heavyblade commented on lib/active_merchant/billing/gateways/sum_up.rb:
```diff
@@ -0,0 +1,192 @@
+      def add_payment(post, payment, options)
+        post[:payment_type] = options[:payment_type] if PAYMENT_TYPES.include?(options[:payment_type])
+        card                = post[:card] = {}
+        card[:name]         = [payment.first_name, payment.last_name].join(' ')
```
💬 Comment:

Maybe just using the `ActiveMerchant::Billing::CreditCard#name` method?

Heavyblade commented on lib/active_merchant/billing/gateways/sum_up.rb:
```diff
@@ -0,0 +1,192 @@
+        end
+
+        # Multiple invalid parameters
+        response = format_multiple_errors(response) if response.is_a?(Array)
```
❓ Question

`response.is_a?(Array)` is the best way to check if the response was successful or not ?

sinourain replied:
You're right, I'm going to change this behavior.

Heavyblade commented on lib/active_merchant/billing/gateways/sum_up.rb:
```diff
@@ -0,0 +1,192 @@
+
+        {
+          error_code: STANDARD_ERROR_CODE_MAPPING[:multiple_invalid_parameters],
+          message: 'Validation error',
```
❓ Question:

Any reason to set here explicitly that all errors are going to be validation errors?

sinourain replied:
Yes, in the [SumUp documentation](https://developer.sumup.com/docs/api/process-a-checkout/) the error 400 (Bad Request) have a response for Multiple_Invalid_Parameters all of them are `Vaidation error`.

Heavyblade replied:
I mean => let's say that you are going to have 401 or 500 errors are they to be shown as "validation errors" any way ?

sinourain replied:
No, according to the documentation there is only one type of Multiple_Invalid_Parameters (Array) so the scope is only for these errors (400 Bad Request)

Heavyblade reviewed (commented):
@sinourain first round Review is done, I'll leave the test review for the last one.

sinourain commented on lib/active_merchant/billing/gateways/sum_up.rb:
```diff
@@ -0,0 +1,192 @@
+module ActiveMerchant #:nodoc:
+  module Billing #:nodoc:
+    class SumUpGateway < Gateway
+      self.live_url = 'https://api.sumup.com/v0.1/'
```
Yes, only one

Heavyblade commented on lib/active_merchant/billing/gateways/sum_up.rb:
```diff
@@ -0,0 +1,198 @@
+        super
+      end
+
+      def purchase(money, payment, options = {})
```
💬 Comment/Refactor:

suggestion if you want:

```ruby
      def purchase(money, payment, options = {})
        MultiResponse.run do |r|
          r.process { create_checkout(money, payment, options) } unless options[:checkout_id].present?
          r.process { complete_checkout(options[:checkout_id] || r.params['id'], payment, options) }
        end
      end

```

Heavyblade commented on lib/active_merchant/billing/gateways/sum_up.rb:
```diff
@@ -0,0 +1,198 @@
+        }
+      end
+
+      def add_credit_card(payment)
```
❓ Question:

What is the difference between the `add_credit_card` method a just calling `clone` on the payment object?

sinourain replied:
It makes sense, I hadn't realized xD

Heavyblade commented on test/remote/gateways/remote_sum_up_test.rb:
```diff
@@ -0,0 +1,112 @@
+    assert_equal 'PENDING', response.params['transactions'].first['status']
+  end
+
+  def test_successful_purchase_with_existing_checkout
```
❓ Question:

Ohhhh this is super interesting my assumption was that for every authorize/purchase transaction would exist just one Checkout (given the checkout definition for SumUp), but it is allowing you to use an "already" used reference for another purchase.

Is that a reasonable use case ?

sinourain replied:
Yes, in SumUp when a checkout is created it can have several transactions, in fact when the checkout is created a payment method is not placed. I will ask...

Heavyblade commented on test/unit/gateways/sum_up_test.rb:
```diff
@@ -0,0 +1,361 @@
+  end
+
+  def test_failed_purchase
+    @credit_card.year = 2020
```
💬 Comment:

GIven that `failed_complete_checkout_response` is fixed changing the year is not going to have any effect

Heavyblade commented on test/unit/gateways/sum_up_test.rb:
```diff
@@ -0,0 +1,361 @@
+    }
+  end
+
+  def test_successful_purchase
```
💬 Comment:

given that response is going to be fixed the only point of having a unit test here is to actually check how the request is being built, so you can confirm that values are in the proper place and format.

Heavyblade commented on test/unit/gateways/sum_up_test.rb:
```diff
@@ -0,0 +1,361 @@
+  def test_scrub
+    assert @gateway.supports_scrubbing?
+    assert_equal @gateway.scrub(pre_scrubbed), post_scrubbed
+  end
```
💬 Comment:

Maybe so small test for some private methods like would ensure that adapter behavior is declared.

- success_from
- message_from
- authorization_from
- format_multiple_errors
- error_code_from

Heavyblade reviewed (commented):
@sinourain done with my final review, thanks for the quick response

naashton reviewed (changes requested):
This all looks really good. I just had a few minor suggestions.

sinourain commented:
> This all looks really good. I just had a few minor suggestions.

@naashton and @Heavyblade thanks for your feedback, I updated the requested changes and also I add an extra change related to the credential: `token`, I changed it to `access_token` in order to prevent collide with a reserved word

naashton commented:
> > This all looks really good. I just had a few minor suggestions.
> 
> @naashton and @Heavyblade thanks for your feedback, I updated the requested changes and also I add an extra change related to the credential: `token`, I changed it to `access_token` in order to prevent collide with a reserved word

Thanks @sinourain - this actually made me curious, why aren't we using the `client_id` and `client_secret` for authentication as the [SumUp docs](https://developer.sumup.com/docs/api/sum-up-rest-api/) describe? Can you share the documentation where using the `pay_to_email` authentication is explained?

sinourain commented:
> > > This all looks really good. I just had a few minor suggestions.
> > 
> > 
> > @naashton and @Heavyblade thanks for your feedback, I updated the requested changes and also I add an extra change related to the credential: `token`, I changed it to `access_token` in order to prevent collide with a reserved word
> 
> Thanks @sinourain, this actually made me curious - Why aren't we using the `client_id` and `client_secret` for authentication as the [SumUp docs](https://developer.sumup.com/docs/api/sum-up-rest-api/) describe? And can you share the documentation where using the `pay_to_email` authentication is explained?

Sure, we discuss it in the Endava team, basically [SumUp have two ways for Authorization, API keys and OAuth 2.0](https://developer.sumup.com/docs/online-payments/introduction/authorization/), We decided to implement API Keys because it made the most sense given that if we implemented OAuth 2.0 it required additional development and requesting permissions to be able to make a purchase.

About the field `pay_to_email` in the [documentation mention it in the Request Body](https://developer.sumup.com/docs/api/create-a-checkout/#create-a-checkout). Curiously, it recommends that we use `merchant_code` instead of the `pay_to_email` field but according to tests we did, it is not possible to omit the `pay_to_email` field.

naashton reviewed (approved):
Awesome work!
